lib/cs_guide/postcode_lat_long.ex

```diff
@@ -0,0 +1,132 @@
+defmodule CsGuide.PostcodeLatLong do
+  alias NimbleCSV.RFC4180, as: CSV
+  alias CsGuide.Resources.Venue
+  import Ecto.Query
+  use GenServer
+
+  @zip_file_name "postcodes.zip"
+  @zip_file_name_charlist String.to_charlist(@zip_file_name)
+  @postcode_file_charlist 'ukpostcodes.csv'
+
+  def start_link do
+    GenServer.start_link(__MODULE__, %{}, name: PostcodeCache)
+  end
+
+  def init(initial_state) do
+    postcode_cache = :ets.new(:postcode_cache, [:set, :protected, :named_table])
+    store_postcodes_in_ets(postcode_cache)
+    {:ok, initial_state}
+  end
+
+  defp store_postcodes_in_ets(ets) do
+    if File.exists?("ukpostcodes.csv") do
+      IO.inspect("Postcode file exists, storing with ets")
+      get_postcodes_from_csv("ukpostcodes.csv", ets)
+    else
+      IO.inspect("Postcode file does not exist. Creating file...")
+      case create_csv_file() do
+        :ok ->
+          IO.inspect("Removed zip file. Storing postcodes with ets")
+          get_postcodes_from_csv("ukpostcodes.csv", ets)
+
+        error ->
+          error
+      end
+    end
+  end
+
+  def nearest_venues(lat, long, distance \\ 5_000)
+  def nearest_venues(lat, long, distance) when distance < 30_000 do
+    case venues_within_distance(distance, lat, long) do
+      [] -> nearest_venues(lat, long, distance + 5_000)
+      venues -> venues
+    end
+  end
+  def nearest_venues(_, _, _), do: []
+
+  defp venues_within_distance(distance, lat, long) do
+    Venue
+    |> where([venue], venue.deleted == false)
+    |> where(
+      [venue],
+      fragment(
```

Comment:
@RobStallion you've done well to split this query out. please consider inline comments to aid maintainability. 👍

Reply:
@nelsonic https://github.com/club-soda/club-soda-guide/pull/291/commits/3c91b6f2db29ec911c0b8020068f803daf2d8965 is this what you had in mind?